torch/csrc/distributed/c10d/ProcessGroup.cpp

```diff
@@ -158,3 +161,137 @@ void ProcessGroup::release_resources() {
 }
 
 } // namespace c10d
+
+namespace {
+
+class WorkRegistry {
+ public:
+  void register_work(
+      const at::Tensor& tensor,
+      const c10::intrusive_ptr<c10d::Work>& work) {
+    if (!tensor.has_storage()) {
+      TORCH_WARN_ONCE(
+          "Registering collective work for tensor without storage is not supported. "
+          "Calling c10d_functional.wait_tensor() on this tensor will not wait for the collective to complete. "
+          "Unsupported tensor type: " +
+          tensor.toString());
+      return;
+    }
+    auto storage = tensor.storage().getWeakStorageImpl();
+    std::unique_lock lock(lock_);
+
+    auto it = registry_.find(storage);
+    if (it == registry_.end()) {
+      registry_.emplace(
+          std::move(storage),
+          std::vector<c10::intrusive_ptr<c10d::Work>>{work});
+    } else {
+      // There is no guarantee that the previous work object for this
+      // tensor storage is completed before the new work object is registered.
+      // Therefore we need to maintain a list of work objects for each tensor
+      // storage.
+      it->second.push_back(work);
+    }
+  }
+
+  std::vector<c10::intrusive_ptr<c10d::Work>> pop_works(
+      const at::Tensor& tensor) {
+    const auto storage = tensor.storage().getWeakStorageImpl();
+    std::unique_lock lock(lock_);
+    auto it = registry_.find(storage);
+    if (it == registry_.end()) {
+      return {};
+    }
+    auto works = it->second;
+    registry_.erase(it);
+    return works;
+  }
+
+  void unregister_completed_works() {
```

Comment:
From my experience fixing the CI errors, this function is usually the culprit. In case we think the CI might not be sufficiently covering all production use cases, we could make this feature into a gradual rollout to minimize potential impact.